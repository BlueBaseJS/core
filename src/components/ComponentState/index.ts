import ComponentState from './ComponentState';
export * from './ComponentState';
export default ComponentState;
