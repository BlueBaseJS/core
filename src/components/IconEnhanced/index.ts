import IconEnhanced from './IconEnhanced';
export default IconEnhanced;
