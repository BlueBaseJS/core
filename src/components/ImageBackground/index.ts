import ImageBackground from './ImageBackground';
export default ImageBackground;
export * from './ImageBackground';
