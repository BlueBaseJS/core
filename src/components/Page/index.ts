import Page from './Page';
export * from './Page';
export default Page;
