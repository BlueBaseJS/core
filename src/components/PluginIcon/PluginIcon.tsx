import { BlueBase } from '../../BlueBase';
import { BlueBaseConsumer } from '../../Context';
import React from 'react';
import { resolveThunk } from '../../utils';

export interface PluginIconProps {
	slug: string;
	size?: number,
	[key: string]: any
}

export class PluginIcon extends React.PureComponent<PluginIconProps> {

	render() {

		return (
			<BlueBaseConsumer>
				{(BB: BlueBase) => {

					const { slug, ...rest } = this.props;

					const plugin = BB.Plugins.get(slug);

					if (!plugin) {
						throw new Error(`There's no pluign registered with "${slug}" key in the registry.`);
					}

					if (!plugin.icon) {
						return null;
					}

					const iconProps = resolveThunk(plugin.icon, plugin, BB);

					return <BB.Components.DynamicIcon {...iconProps} {...rest} />;
				}}
			</BlueBaseConsumer>
		);
	}
}
