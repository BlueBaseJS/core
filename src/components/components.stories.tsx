import * as React from 'react';
// tslint:disable-next-line:no-implicit-dependencies
import { storiesOf } from '@storybook/react';
import BR from '../index';
import ComponentStateButton from '../components/ComponentState/ComponentStateButton';

const stories = storiesOf('Avatar', module);

stories.add('ComponentStateButton', () => {
	return (
    <ComponentStateButton/>);
});
