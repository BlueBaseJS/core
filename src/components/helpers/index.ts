import react from 'react';
const isHidden = (screenSize, props) => {
	switch (screenSize) {
		case 'sm':
			return !!props.smHidden;
		case 'md':
			return !!props.mdHidden;
		case 'lg':
			return !!props.lgHidden;
		case 'xl':
			return !!props.xlHidden;
		case 'xs':
			return !!props.xsHidden;
		default:
			return false;
	}
};

const toPercent = num => `${num * 100}%`;

const getComponentWidth = (screenSize, props) => {
	switch (screenSize) {
		case 'sm':
			if (props.sm) {
				return toPercent(props.sm / props.rowSize);
			}
			return props.parentWidth;

		case 'xs':
			if (props.xs) {
				return toPercent(props.xs / props.rowSize);
			}
			return props.parentWidth;

		case 'xl':
			if (props.xl) {
				return toPercent(props.xl / props.rowSize);
			}
			return props.parentWidth;

		case 'md':
			if (props.md) {
				return toPercent(props.md / props.rowSize);
			} else if (props.sm) {
				return toPercent(props.sm / props.rowSize);
			}
			return props.parentWidth;

		case 'lg':
			if (props.lg) {
				return toPercent(props.lg / props.rowSize);
			} else if (props.md) {
				return toPercent(props.md / props.rowSize);
			} else if (props.sm) {
				return toPercent(props.sm / props.rowSize);
			}
			return props.parentWidth;

		default:
			return props.parentWidth;
	}
};

const getComponentOffset = (screenSize, props) => {
	switch (screenSize) {
		case 'sm':
			if (props.smOffset) {
				return toPercent(props.smOffset / props.rowSize);
			}
			return 0;

		case 'xs':
			if (props.xsOffset) {
				return toPercent(props.xsOffset / props.rowSize);
			}
			return 0;

		case 'xl':
			if (props.xlOffset) {
				return toPercent(props.xlOffset / props.rowSize);
			}
			return 0;

		case 'md':
			if (props.mdOffset) {
				return toPercent(props.mdOffset / props.rowSize);
			} else if (props.smOffset) {
				return toPercent(props.smOffset / props.rowSize);
			}
			return 0;

		case 'lg':
			if (props.lgOffset) {
				return toPercent(props.lgOffset / props.rowSize);
			} else if (props.mdOffset) {
				return toPercent(props.mdOffset / props.rowSize);
			} else if (props.smOffset) {
				return toPercent(props.smOffset / props.rowSize);
			}
			return 0;
		default:
			return 0;
	}
};

export { isHidden, getComponentWidth, getComponentOffset };
