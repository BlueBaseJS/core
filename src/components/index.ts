export * from './BlueBaseApp';
export * from './BlueBaseContent';
export * from './BlueBaseFilter';
export * from './ComponentState';
export * from './DataObserver';
export * from './DynamicIcon';
export * from './EmptyState';
export * from './ErrorObserver';
export * from './ErrorState';
export * from './Header';
export * from './HeaderBackButton';
export * from './HeaderTitle';
export * from './HomeScreen';
export * from './HoverObserver';
export * from './JsonSchema';
export * from './Link';
export * from './LoadingState';
export * from './Navigation';
export * from './NavigationActions';
export * from './Noop';
export * from './PluginIcon';
export * from './ReactLoadableLoading';
export * from './Redirect';
export * from './StatefulComponent';
export * from './WaitObserver';
