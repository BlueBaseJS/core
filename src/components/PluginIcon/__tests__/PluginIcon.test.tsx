import { BlueBaseApp } from '../../../';
import { DynamicIconProps } from '@bluebase/components';
import { PluginIcon } from '../PluginIcon';
import React from 'react';
import TestRenderer from 'react-test-renderer';
import { Text } from 'react-native';


describe('PluginIcon', () => {

	// it(`Snapshot PluginIcon component with no plugin registered`, () => {

	// 	const component = TestRenderer.create(
	// 		<BlueBaseApp>
	// 			<PluginIcon id="unregistered-plugin" />
	// 		</BlueBaseApp>
	// 	);
	// 	try {
	// 		const tree = component.toJSON();
	// 		expect(tree).toMatchSnapshot();
	// 	} catch (e) {
	// 		expect(e.message).toBe(`There's no pluign registered with "unregistered-plugin" key in the registry.`);
	// 	}
	// });

	test(`should render an image icon for a registered plugin`, (done) => {

		const plugin = {
			icon: {
				source: { uri: 'https://picsum.photos/200' },
				type: 'image' as DynamicIconProps['type'],
			},
			key: 'some',
			value: {},
		};

		const rendered = TestRenderer.create(
			<BlueBaseApp plugins={[plugin]}>
				<PluginIcon id="some" />
			</BlueBaseApp>
		);

		setTimeout(() => {
			const tree = rendered.toJSON();
			// expect(tree).toMatchSnapshot();

			expect((tree as any).type).toBe('Image');
			expect((tree as any).props.size).toBe(100);
			expect((tree as any).props.style).toMatchObject({ height: 100, width: 100 });
			done();
		});
	});

	test(`should render an image icon for a registered plugin where icon prop is a thunk`, (done) => {

		const plugin = {
			icon: () => ({
				source: { uri: 'https://picsum.photos/200' },
				type: 'image' as DynamicIconProps['type'],
			}),
			key: 'some',
			value: {},
		};

		const rendered = TestRenderer.create(
			<BlueBaseApp plugins={[plugin]}>
				<PluginIcon id="some" />
			</BlueBaseApp>
		);

		setTimeout(() => {
			const tree = rendered.toJSON();
			// expect(tree).toMatchSnapshot();

			expect((tree as any).type).toBe('Image');
			expect((tree as any).props.size).toBe(100);
			expect((tree as any).props.style).toMatchObject({ height: 100, width: 100 });
			done();
		});
	});

	test(`should render null where plugin doesnt have an icon prop`, (done) => {

		const plugin = {
			key: 'some',
			value: {},
		};

		const rendered = TestRenderer.create(
			<BlueBaseApp plugins={[plugin]}>
				<PluginIcon id="some" />
			</BlueBaseApp>
		);

		setTimeout(() => {
			const tree = rendered.toJSON();
			// expect(tree).toMatchSnapshot();

			expect((tree as any)).toBe(null);
			done();
		});
	});

	test(`should throw an error when a plugin is not registered`, (done) => {

		const rendered = TestRenderer.create(
			<BlueBaseApp>
				<PluginIcon id="some" />
			</BlueBaseApp>
		);

		setTimeout(() => {
			const found = rendered.root.findAllByType(Text);
			expect(
				(found[1].children[0] as any).children.join()
			).toBe('There\'s no pluign registered with "some" key in the registry.');

			done();
		});
	});

});
