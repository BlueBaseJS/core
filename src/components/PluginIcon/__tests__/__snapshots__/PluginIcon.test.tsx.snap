// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginIcon Snapshot PluginIcon component with no plugin registered 1`] = `
<Text>
  Loading
</Text>
`;

exports[`PluginIcon should render an image icon for a registered plugin 1`] = `
<View>
  <Image
    size={100}
    source={
      Object {
        "uri": "https://picsum.photos/200",
      }
    }
    style={
      Object {
        "height": 100,
        "width": 100,
      }
    }
  />
</View>
`;

exports[`PluginIcon should render an image icon for a registered plugin where icon prop is a thunk 1`] = `
<View>
  <Image
    size={100}
    source={
      Object {
        "uri": "https://picsum.photos/200",
      }
    }
    style={
      Object {
        "height": 100,
        "width": 100,
      }
    }
  />
</View>
`;

exports[`PluginIcon should render null where plugin doesnt have an icon prop 1`] = `<View />`;
